### ISSUE TITLE
Connection pool fails if server closes connection

### ISSUE THREAD

**Issue by FauxFaux** [algesten/ureq#10]
The connection pool does not cope with HTTPS servers closing the connection, due to `rustls` being mean.

If I make a request, then wait for the connection to time out, then make another request, I see the error: `BadStatus`. This means that every other request I make fails.

This test fails, because https://fau.xxx/ is currently running an nginx with a `keepalive_timeout` of 2s:

```
#[test]
fn connection_reuse() {
    let agent = ureq::Agent::default().build();
    let resp = agent.get("https://fau.xxx/").call();

    // use up the connection so it gets returned to the pool
    assert_eq!(resp.status(), 200);
    resp.into_reader().read_to_end(&mut vec![]).unwrap();

    // wait for the server to close the connection
    std::thread::sleep(Duration::from_secs(3));

    // try and make a new request on the pool
    let resp = agent.get("https://fau.xxx/").call();
    if let Some(err) = resp.synthetic_error() {
        panic!("boom! {:?}", err);
    }
    assert_eq!(resp.status(), 200);
}
```

nginx defaults to 75s, some servers have much longer timeouts, some much shorter, but everyone will eventually see this problem.

I was hoping that attempting a write during `send_prelude` would trigger the retry code, #8, but this does not help.

I do not see a way to fix this right now. A `read(&mut [])?` in `send_prelude` doesn't trigger it, and we aren't expecting any data to be readable at that point, so clever buffering wouldn't help.

**Comment by lolgesten** [comment]
The connection pool overall is a bit under tested. I'm all open to suggestions here.

**Comment by algesten** [comment]
I made a fix that goes somewhat towards a solution. See commit message.

**Comment by jsha** [comment]
Additional fixes landed in 1.3.0, and I think we can consider this done now.
